### PR TITLE
Trama

### PR DESCRIPTION
**Pull request by bkamuz**
# Source: Tradingview "Trend Regularity Adaptive Moving Average"

## Pinescript Author: LuxAlgo

Hello! I decided to discard all previous changes and create a new pull-request for the TRAMA indicator. Here is an improved version. Failed to make it fully vectorized (I’m sure there is an approach but unfortunately didn't get to it yet). This version is quite fast, compared to other versions I’ve tried.

[Trend Regularity Adaptive Moving Average - LuxAlgo](https://tradingview.com/script/p8wGCPi6-Trend-Regularity-Adaptive-Moving-Average-LuxAlgo/)

Screenshot from 2024-09-23 23-09-08

Screenshot from 2024-09-23 23-10-59

### CODE REVIEW

File: technical/indicators/indicators.py [hunk 1]
```diff
@@ -1353,3 +1353,38 @@ def tv_alma(
 
     alma = np.convolve(dataframe[field], weights[::-1], mode="valid")
     return Series(np.pad(alma, (length - 1, 0), mode="constant", constant_values=np.nan))
+
+
+def tv_trama(dataframe: DataFrame, length: int = 99, field="close") -> DataFrame:
```
Comment (xmatthias) [inline technical/indicators/indicators.py:1358]: The actual response type is Series - not dataframe.

File: technical/indicators/indicators.py [hunk 2]
```diff
@@ -1353,3 +1353,38 @@ def tv_alma(
...
+
+    hh = ta.MAX(dataframe["high"], length)
+    ll = ta.MIN(dataframe["low"], length)
+    hh_or_ll = np.where(np.diff(hh) > 0, 1, 0) + np.where(np.diff(ll) < 0, 1, 0)
+
+    tc = np.zeros(df_len)
+    tc[:-1] = np.nan_to_num(ta.SMA(hh_or_ll.astype(float), length) ** 2)
+
+    ama = np.zeros(df_len)
+    ama[0] = dataframe[field][0]
```
Comment (xmatthias) [inline technical/indicators/indicators.py:1386]: it's not quite clear what you're trying to do - but this is quite error-prone as it assumes that an index with "0" exists.

This is not necessarily the case and not something we should rely on (a user could've moved the date column to the index, removed starting lines, ...).
If you really need the very first value of the list - please use `.iloc[0]` instead.

obviously the same applies to the below line.

You can test this pretty simply by passing `dataframe[1000:]` into the function (with a dataframe that has more than 1000 lines).

**Review by xmatthias — CHANGES_REQUESTED** [review]
In it's current state, it's crashing in quite common scenarios (keyerror due to the assumption that an index starts at 0 and has no missing lines).

File: technical/indicators/indicators.py (hunk 2 above)
Comment (bkamuz) [reply technical/indicators/indicators.py:1386]: Got it. Thanks for the help.

**Review by xmatthias — APPROVED** [review]
Great job - :+1:

**Pull request merged**